pkg/queue/extensions.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"context"
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// QPExtension is the interface for interacting with queue proxy injection
+// implementations. Extending this interface will allow adding more
+// QP extendiability.
+type QPExtension interface {
+	Init(logger *zap.SugaredLogger, ctx context.Context)
+	Shutdown()
+
+	// If extension does not require to be added to Transport
+	// (e.g. when the extensoin is not active),
+	// Transport should return next (never return nil)
+	Transport(next http.RoundTripper) (roundTripper http.RoundTripper)
+}
```

Comment:
1. **To Interface or not to Interface, that is the question:** 
If we interface:
         (a) we open up for growing extendibility as we move forward, without cluttering the core code. 
         (b) we are not bound to limit ourselves to a single call to the extension (and maneuvering around it)
         (c) in the future, if we allow several extension points, we can allow an extension to use more than one 
         (d) this requires extensions to align with the interface (once it changes to include additional calls)
     
     If we do seek to squeeze into a single extendiability point, we can have:
              ` e.transport = Transport(ctx, logger, transport)`
     This will work for removing `e.Init(ctx, logger)`, but is this the right way forward?
     
      It is a reasonable assumption that other extendibility will be needed as we move forward, and all will need `Init()`. For example, we may have an extension that only requires `Init()` but not `Transport()` (Do we really want it to use `Transport()` as a fake `Init()`, making the code unreadable?). Also, we may have an extension (security, tracing, metrics, container freeze, etc. ) requiring additional hooks beyond `Transport()`. Instead of creating a MainWithThisHook() and MainWIthThatHook() for each one, we can make it generic and all will use the same interface (and use what they need from that interface, skipping the rest).      
      
      As for the `e.Shutdown()`, if we avoid the interface, we can solve it in `main()`. 
      But if we keep the interface, having it seems like the right way to go.
      

1. 
(a) We need: **One logger to rule them all**. 
      Extensions need to use the logger controlled and being setup by core. 
     
```
logger, _ := pkglogging.NewLogger(env.ServingLoggingConfig, env.ServingLoggingLevel)
defer flush(logger)

logger = logger.Named("queueproxy").With(
	zap.String(logkey.Key, types.NamespacedName{
		Namespace: env.ServingNamespace,
		Name:      env.ServingRevision,
	}.String()),
	zap.String(logkey.Pod, env.ServingPod)) 
```
      
       
(b) Right, the order needs to change to be `Init(ctx, logger)`